Modify tag helper infrastructure to configure case sensitivity

ajaybhargavb opened the pull request:
Part of https://github.com/aspnet/AspNetCore/issues/9860

- Added `CaseSensitive` property to all tag helper descriptors
- Flow that property through to use that to match tag helpers
- Updated Component tag helper descriptors to be case sensitive
- Added some sample tests

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/AllowedChildTagDescriptor.cs:
```diff
@@ -13,6 +13,8 @@ public abstract class AllowedChildTagDescriptor : IEquatable<AllowedChildTagDesc
 
         public string DisplayName { get; protected set; }
 
+        public bool CaseSensitive { get; protected set; }
```
I assume this breaking change is ok.

rynowak replied:
What's breaking about it?

ajaybhargavb replied:
I meant to add that comment on the builder. Ignore

ajaybhargavb commented:
This is the experience after this change,
![repro](https://user-images.githubusercontent.com/1579269/60934091-ccaad100-a279-11e9-9c9e-0493dfc05051.gif)

It is as expected (even completions) except for some odd issues with coloring. Changes from WTE should fix this.

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/AllowedChildTagDescriptor.cs:
```diff
@@ -32,6 +34,11 @@ public override string ToString()
+            if (CaseSensitive)
+            {
+                return AllowedChildTagDescriptorComparer.CaseSensitive.Equals(this, other);
+            }
```
I'm really not a fan of having *a special comparer* and we should nuke it from orbit. Specifically, the idea of this is wrong, it's a comparer that compares **EVERYTHING** about the tag helper as case sensitive.

IMO equality should behave in the way its currently defined, and the binder should be the thing that enforces case-sensitivity.

Also, this is really problematic because it's non-reflexive.

ajaybhargavb replied:
> Specifically, the idea of this is wrong, it's a comparer that compares EVERYTHING about the tag helper as case sensitive.

That is not the case though. It compares as case sensitive only for properties that makes sense. See https://github.com/aspnet/AspNetCore-Tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.Language/BoundAttributeDescriptorComparer.cs#L54 for example.

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/AllowedChildTagDescriptorBuilder.cs:
```diff
@@ -11,6 +11,8 @@ public abstract class AllowedChildTagDescriptorBuilder
 
         public abstract string DisplayName { get; set; }
 
+        public abstract bool CaseSensitive { get; set; }
```
This is a breaking change that can easily be avoided.

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/TagHelperDescriptorBuilder.cs:
```diff
@@ -52,6 +52,8 @@ public static TagHelperDescriptorBuilder Create(string kind, string name, string
 
         public abstract string TagOutputHint { get; set; }
 
+        public abstract bool CaseSensitive { get; set; }
```
I think there's probably too many levels of case sensitivity here - or at least more than we know we need. What would you think about applying case sensitivity at the top level (tag helper).

NTaylorMullen replied:
I had the same initial reaction. However, looking through our code base if we were to remove the CaseSensitive properties on the descriptors we'd need to flow that information independently through a good portion of our APIs. For instance, as one example take a look at the `TagHelperMatchingConventions` and all of the independent methods that inspect just the `BoundAttributeDescriptor` or just the `TagMatchingRuleDescriptor`. 

The core requirement of this feature is three pronged, in regards to enforcing case sensitivity. We need to enforce it for:
1. Required attributes
2. Bound Attributes
3. Required Tag Names

Given those three requirements we need to handle them in some fashion whether we pass additional "case sensitivity = true" parameters or have them be on the descriptors. Oh and, we could nuke the property on `AllowedChildDescriptor` since that concept doesn't exist in Component world.

I don't love what's here currently but I think @ajaybhargavb had to pick between the better of two evils. Let's use our Razor sync up today to see if we can come up with something better or agree on what's here.

ajaybhargavb replied:
@NTaylorMullen described exactly why I took this approach.
In addition,
- I realize `AllowedChildDescriptor` doesn't make sense for components. But I did it for consistency and we won't be able to change it once we ship 3.0. So I just wanted to make sure we had this capability.
- Another advantage of adding `CaseSensitive` on individual required attributes/bound attribute/tag name etc is that it allows us to customize the descriptors to make it so that some are case sensitive while others aren't. I know we don't really need this now but I see no harm in having that.

> Let's use our Razor sync up today to see if we can come up with something better or agree on what's here.

👍

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/TagHelperMatchingConventions.cs:
```diff
@@ -229,13 +230,13 @@ internal static bool SatisfiesRequiredAttribute(string attributeName, string att
                 // attributeName cannot equal the Name if comparing as a PrefixMatch.
                 nameMatches = attributeName.Length != descriptor.Name.Length &&
-                    attributeName.StartsWith(descriptor.Name, StringComparison.OrdinalIgnoreCase);
+                    attributeName.StartsWith(descriptor.Name, descriptor.CaseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase);
```
This is the good part of the change 👍

ajaybhargavb commented:
🆙 📅 No longer in draft
- Nuked the case sensitive comparer
- Added more tests

ajaybhargavb commented:
I'll add diagnostics in the next PR.

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/AllowedChildTagDescriptorComparer.cs:
```diff
@@ -55,15 +35,16 @@ private AllowedChildTagDescriptorComparer(bool caseSensitive = false)
             return
-                string.Equals(descriptorX.Name, descriptorY.Name, _stringComparison) &&
+                descriptorX.CaseSensitive == descriptorY.CaseSensitive &&
+                string.Equals(descriptorX.Name, descriptorY.Name, descriptorX.CaseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase) &&
```
So it should always be Ordinal? I just maintained the behavior of the CaseSensitive comparer

NTaylorMullen replied:
Yup, we're just comparing descriptor objects so we can do things like detect if we need to re-compute TagHelpers.

ajaybhargavb replied:
I just checked and it looks like this isn't the descriptor name but the actual allowed child tag name.

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/BoundAttributeDescriptorComparer.cs:
```diff
@@ -50,13 +31,21 @@ public virtual bool Equals(BoundAttributeDescriptor descriptorX, BoundAttributeD
+                return false;
+            }
+
+            var stringComparer = descriptorX.CaseSensitive ? StringComparer.Ordinal : StringComparer.OrdinalIgnoreCase;
```
Ditto for this. `descriptor.Name` is the attribute name.

ajaybhargavb commented:
🆙 📅 

- Made `CaseSensitive` settable only at the top-level
- Added a diagnostic when component file names start with a lowercase letter
- Added a diagnostic(warning) when a markup element does not start with an lowercase letter

Note:
I haven't addressed a case when you do `<Counter></counter>`. This is currently allowed. The fix for this isn't as easy as I thought as it involves special casing the parser. I am leaning towards simply adding a diagnostic when a component has casing mismatch in start and end tags. That will be simple and would avoid complicating the parser. Thoughts?

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDiagnosticFactory.cs:
```diff
@@ -421,5 +421,33 @@ public static RazorDiagnostic Create_RefSuppressFieldRequiredForGeneric(SourceSp
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}10014",
+                () => "Found markup element with unexpected name '{0}'. If this is intended to be a component, add a @using directive for its namespace.",
+                RazorDiagnosticSeverity.Warning);
```
Since Todd is back now does he know about our ask for warnings? I know we met with WTE but wanted to make sure that one requirement was still being tracked on their end.

ajaybhargavb replied:
Yeah. I mentioned it in our meeting and in the email I sent

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/RequiredAttributeDescriptorComparer.cs:
```diff
@@ -58,18 +39,24 @@ private RequiredAttributeDescriptorComparer(bool caseSensitive = false)
+
             var hash = HashCodeCombiner.Start();
-            hash.Add(descriptor.Name, _stringComparer);
+            hash.Add(descriptor.Name, descriptor.CaseSensitive ? StringComparer.Ordinal : StringComparer.OrdinalIgnoreCase);
```
Missed this guy 😄

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Serialization/TagHelperDescriptorJsonConverter.cs:
```diff
@@ -34,13 +34,15 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             var childTags = descriptor[nameof(TagHelperDescriptor.AllowedChildTags)].Value<JArray>();
             var documentation = descriptor[nameof(TagHelperDescriptor.Documentation)].Value<string>();
             var tagOutputHint = descriptor[nameof(TagHelperDescriptor.TagOutputHint)].Value<string>();
+            var caseSensitive = descriptor[nameof(TagHelperDescriptor.CaseSensitive)].Value<bool>();
```
Would you mind logging an issue in Razor.VSCode and putting it in the next release milestone to update its json converter with these changes?

ajaybhargavb replied:
Filed https://github.com/aspnet/Razor.VSCode/issues/377. Could't find any relevant milestones so left if empty.

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDocumentClassifierPass.cs:
```diff
@@ -64,6 +64,14 @@ protected override CodeTarget CreateTarget(RazorCodeDocument codeDocument, Razor
             }
 
+            var documentNode = codeDocument.GetDocumentIntermediateNode();
+            if (char.IsLower(computedClass, 0))
```
I'd double check the implications of having the error logged at this layer. For instance at build time I believe you'll get two duplicate errors. one from the declaration generation pass and one from the actual razor compile.

It might make more sense for you to do this check at the TagHelperDescriptor provider layer. The descriptors have their own diagnostics which would also make sense to put this there.

ajaybhargavb replied:
That doesn't happen when I test it locally. Also, we can't do it at the provider layer because we don't have access to the CodeDocument->SourceDocument->FilePath->FileName. I need that to provide this diagnostic. So, it doesn't make sense to put it there.

NTaylorMullen replied:
If it doesn't happen then something else must be preventing it. Alright, can ignore this then 😄

NTaylorMullen commented on src/Razor/test/RazorLanguage.Test/IntegrationTests/ComponentCodeGenerationTestBase.cs:
```diff
@@ -4429,6 +4515,30 @@ public void DuplicateMarkupAttributes_IsAnError_BindValue()
+}
+");
+
+
```
🔥

NTaylorMullen commented:
> I haven't addressed a case when you do `<Counter></counter>`. This is currently allowed. The fix for this isn't as easy as I thought as it involves special casing the parser. I am leaning towards simply adding a diagnostic when a component has casing mismatch in start and end tags. That will be simple and would avoid complicating the parser. Thoughts?

Hmm, why in the parser? If it's because the parser groups HTML elements together despite case, can't we pull them apart?

NTaylorMullen commented on src/Razor/test/RazorLanguage.Test/IntegrationTests/ComponentCodeGenerationTestBase.cs:
```diff
@@ -4870,6 +4980,35 @@ public class Counter : ComponentBase
+{
+    public class MyComponent : ComponentBase
+    {
+        [Parameter] int IntProperty { get; set; }
```
public on all parameters in the PR

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDocumentClassifierPass.cs:
```diff
@@ -64,6 +64,14 @@ protected override CodeTarget CreateTarget(RazorCodeDocument codeDocument, Razor
+            {
+                // We don't allow component names to start with a lowercase character.
+                documentNode.Diagnostics.Add(
+                    ComponentDiagnosticFactory.Create_ComponentNamesCannotStartWithLowerCase(computedClass, documentNode.Source));
```
That is not what happened when I test it. It only shows up in the ErrorList without any squiggly

NTaylorMullen replied:
Oh interesting document node must not have a source?

ajaybhargavb commented:
> Hmm, why in the parser? If it's because the parser groups HTML elements together despite case, can't we pull them apart?

Yeah. One other option is to special case components and group tags case sensitively in the parser. The tag helper rewriting system was specifically designed to be dumb and follow the grouping done by the parser to avoid further complication. I don't want to change that for this. I think adding a diagnostic is a pretty good alternative to doing all that work and complicating things.

NTaylorMullen commented:
> I think adding a diagnostic is a pretty good alternative to doing all that work and complicating things.

Ya I think that's fine, get buy off from @rynowak on that though.

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorIntermediateNodeLoweringPhase.cs:
```diff
@@ -1184,6 +1184,18 @@ public override void VisitMarkupElement(MarkupElementSyntax node)
+                if (node.StartTag != null)
+                {
+                    var startTagName = node.StartTag.GetTagNameWithOptionalBang();
+                    if (startTagName != null && startTagName.Length > 0 && !char.IsLower(startTagName, 0))
```
Yeah I think `char.IsUpper` makes sense. For cases like, `<!Component>` we don't want to provide a warning because we already know this was not intended to be a component because of the bang.

NTaylorMullen replied:
> For cases like, `<!Component>` we don't want to provide a warning because we already know this was not intended to be a component because of the bang.

Yup, totally agree. Should make sure how char.IsUpper handles `!` to ensure it doesn't create the diagnostic is all.

ajaybhargavb replied:
> Each time something in your .razor source resolves as an element (not a component), check that its name starts with a lowercase character in a-z. If we find an element whose first char isn't in a-z case sensitively, emit a warning

This was in the issue description given by @rynowak. @rynowak was this wording intentional? Would we changing the behavior if we use `char.IsUpper` instead of `!char.IsLower`?

ajaybhargavb replied:
> Should make sure how char.IsUpper handles ! to ensure it doesn't create the diagnostic is all.

It returns `false` for `!`. We're good.

ajaybhargavb commented:
> A component usage with two parameters that only differentiate in case. <MyComponent Parameter1="..." PARAMeter2="..." />
Looks like we don't support this even though we can. See https://github.com/aspnet/AspNetCore-Tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDiagnosticFactory.cs#L378

NTaylorMullen commented:
> Looks like we don't support this even though we can. See https://github.com/aspnet/AspNetCore-Tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDiagnosticFactory.cs#L378

Very interesting, @rynowak is this a scenario you want to support?

ajaybhargavb commented:
🆙 📅 
- Added another diagnostic for `<Counter></counter>` case
- Fixed/Added tests
- Addressed feedback

ajaybhargavb commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorIntermediateNodeLoweringPhase.cs:
```diff
@@ -1184,6 +1184,19 @@ public override void VisitMarkupElement(MarkupElementSyntax node)
                             BuildSourceSpanFromNode(node.EndTag), node.EndTag.GetTagNameWithOptionalBang()));
                 }
 
+                if (node.StartTag != null && !_document.Options.SuppressPrimaryMethodBody)
```
Without the second condition here, some tests fail. Because some elements are not recognized as tag helper until we get to the second phase of two phase compilation. So this warning gets added during the first phase and breaks the test. Is there are better way to fix it?

rynowak replied:
Hmmm, I think I'd have to understand how/what we're testing in this case. During the declaration phase we wipe away the method body (including the markup). So I guess lets look a the test strategy being used and try to figure this out.

rynowak replied:
Ah I see now that you're checking the condition. I think this is a totally sensible solution - we genuinely don't want this error during the declaration pass 👍

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDiagnosticFactory.cs:
```diff
@@ -421,5 +421,48 @@ public static RazorDiagnostic Create_RefSuppressFieldRequiredForGeneric(SourceSp
+        public static readonly RazorDiagnosticDescriptor UnexpectedMarkupElement =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}10014",
+                () => "Found markup element with unexpected name '{0}'. If this is intended to be a component, add a @using directive for its namespace.",
```
What do they do if it isn't intended to be a component. Does using `!` help?

ajaybhargavb replied:
Yes

rynowak commented on src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDiagnosticFactory.cs:
```diff
@@ -421,5 +421,48 @@ public static RazorDiagnostic Create_RefSuppressFieldRequiredForGeneric(SourceSp
+        public static readonly RazorDiagnosticDescriptor InconsistentStartAndEndTagName =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}10015",
+                () => "The start tag name '{0}' does not match the end tag name '{1}'. Components must have matching start and end tag names (case-sensitive).",
```
<3

rynowak commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Serialization/TagHelperDescriptorJsonConverter.cs:
```diff
@@ -34,13 +34,15 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
 
             builder.Documentation = documentation;
             builder.TagOutputHint = tagOutputHint;
+            builder.CaseSensitive = caseSensitive;
```
It's nice that we use the builder here, or else we'd have to update all of these serializers 😆

rynowak commented on src/Razor/test/RazorLanguage.Test/IntegrationTests/ComponentCodeGenerationTestBase.cs:
```diff
@@ -998,6 +998,24 @@ public void BuiltIn_BindToInputWithoutType_WritesAttributes()
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
```
Should this have a diagnostic? DO we end up treating it as an expression in this case?

ajaybhargavb replied:
Yeah we just treat it like a regular attributes with an expression value. I'm not sure if we want to have a diagnostic every time a tag helper descriptor could match but can't because of casing.

rynowak commented on src/Razor/test/RazorLanguage.Test/IntegrationTests/ComponentCodeGenerationTestBase.cs:
```diff
@@ -1828,12 +1846,12 @@ public class SomeComponent : ComponentBase
             AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
             AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
-            var result = CompileToAssembly(generated);
+            CompileToAssembly(generated, throwOnFailure: false);
```
Validate the diagnostic?

ajaybhargavb replied:
It is already validated by testname.diagnostics.txt

rynowak commented on src/Razor/test/RazorLanguage.Test/IntegrationTests/ComponentCodeGenerationTestBase.cs:
```diff
@@ -4429,6 +4515,29 @@ public void DuplicateMarkupAttributes_IsAnError_BindValue()
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+
+            var diagnostic = Assert.Single(generated.Diagnostics);
+            Assert.Same(ComponentDiagnosticFactory.DuplicateMarkupAttributeDirective.Id, diagnostic.Id);
```
<3

rynowak commented on src/Razor/test/RazorLanguage.Test/IntegrationTests/ComponentDiagnosticRazorIntegrationTest.cs:
```diff
@@ -125,5 +127,68 @@ public void DirectiveAttribute_ComplexContent_ReportsError()
+        {
+            // Arrange & Act
+            var generated = CompileToCSharp(@"
+<PossibleComponent></PossibleComponent>
```
Can we test that `!` can be used to work around this?

ajaybhargavb replied:
Added a test.

rynowak commented on src/Razor/test/RazorLanguage.Test/IntegrationTests/ComponentTemplateIntegrationTest.cs:
```diff
@@ -58,8 +58,10 @@ public class MyComponent : ComponentBase
-            Assert.Equal("RZ1005", diagnostic.Id);
+            Assert.Collection(
+                generated.Diagnostics,
+                d => Assert.Equal("RZ10014", d.Id),
```
hmmm whatsit?

ajaybhargavb replied:
Ah yes. This was because `UseTwoPhaseCompilation` was set to `false`. I'll change it.

ajaybhargavb commented:
> Very interesting, @rynowak is this a scenario you want to support?

@NTaylorMullen, @rynowak confirmed offline that we don't want to support component usage with two parameters that only differentiate in case.

ajaybhargavb commented:
🆙 📅

ajaybhargavb commented:
Merging now. @NTaylorMullen if you have more feedback, I'll address in a separate PR.

NTaylorMullen commented:
> Merging now. @NTaylorMullen if you have more feedback, I'll address in a separate PR.

Looks good tome
